# Практическое задание 5.
## КМБО-01-20, Гребнев Никита, Большаков Михаил.
## Тема №23 "Употребление алкоголя учащимися".

### Устанавливаем необходимые библиотеки (если они ранее не были установлены):

### Считываем данныеиз файлов:

In [1]:
import pandas as pd
import numpy as np
df1 = pd.read_csv('student-por.csv')
df2 = pd.read_csv('student-mat.csv')

In [2]:
df1

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
644     MS   F   19       R     GT3       T     2     3  services     other   
645     MS   F   18       U     LE3       T     3     1   teacher  services   
646     MS   F   18       U     GT3       T     1     1     other     other   
647     MS   M   17       U     LE3       T     3     1  services  services   
648     MS   M   18       R     LE3       T     3     2  services     other   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        4   0  11  11  
1    ...      5        3      3     1     1      3        2   9  11  11  
2    ...      4        3      2     2     3      3        6  12  13  12  
3    ...      3        2      2     1     1      5        0  14  14  14  
4    ...      4        3      2     1     2      5        0  11  13  13  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
644  ...      5        4      2     1     2      5        4  10  11  10  
645  ...      4        3      4     1     1      1        4  15  15  16  
646  ...      1        1      1     1     1      5        6  11  12   9  
647  ...      2        4      5     3     4      2        6  10  10  10  
648  ...      4        4      1     3     4      5        4  10  11  11  

[649 rows x 33 columns]

In [3]:
df2

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

### Поскольку в df1(курс португальского языка) и df2(курс математики) одинаковые атрибуты, можно объединить датасеты в один:

In [4]:
df = df1.append(df2, ignore_index=True)
# Либо можно так, но тогда индексы второго датасета будут идти с нуля, а не с 396: 
#frames = [df1, df2]
#df = pd.concat(frames)

In [5]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0        GP   F   18       U     GT3       A     4     4   at_home   teacher   
1        GP   F   17       U     GT3       T     1     1   at_home     other   
2        GP   F   15       U     LE3       T     1     1   at_home     other   
3        GP   F   15       U     GT3       T     4     2    health  services   
4        GP   F   16       U     GT3       T     3     3     other     other   
...     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
1039     MS   M   20       U     LE3       A     2     2  services  services   
1040     MS   M   17       U     LE3       T     3     1  services  services   
1041     MS   M   21       R     GT3       T     1     1     other     other   
1042     MS   M   18       R     LE3       T     3     2  services     other   
1043     MS   M   19       U     LE3       T     1     1     other   at_home   

      ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0     ...      4        3      4     1     1      3        4   0  11  11  
1     ...      5        3      3     1     1      3        2   9  11  11  
2     ...      4        3      2     2     3      3        6  12  13  12  
3     ...      3        2      2     1     1      5        0  14  14  14  
4     ...      4        3      2     1     2      5        0  11  13  13  
...   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
1039  ...      5        5      4     4     5      4       11   9   9   9  
1040  ...      2        4      5     3     4      2        3  14  16  16  
1041  ...      5        5      3     3     3      3        3  10   8   7  
1042  ...      4        4      1     3     4      5        0  11  12  10  
1043  ...      3        2      3     3     3      5        5   8   9   9  

[1044 rows x 33 columns]

In [6]:
# Вывод всех строк:
#with pd.option_context('display.max_rows', 2000, 'display.max_columns', 10):  
    #print(df)

## 1. Сколько в датасете объектов и признаков? Дать описание каждому признаку, если оно есть.
---

Размер таблицы - `1044 rows × 33 columns`, из чего можно сделать вывод, что в датасете **1044 объекта и 33 признака (атрибута)**.

**Описание признаков:**

   1. school - школа ученика (бинарное значение):
       - 'GP' - Gabriel Pereira 
       - 'MS' - Mousinho da Silveira
   1. sex - пол ученика (бинарное значение):
       - 'F' - female - женский
       - 'M' - male - мужской
   1. age - возраст ученика (числовое значение от 15 до 22)
   1. address - Тип населённого пункта (бинарное значение):
       - 'U' - urban - городской
       - 'R' - rural - сельский
   1. famsize - размер семьи (бинарное значение): 
       - 'LE3' - $\leq$ 3 человек
       - 'GT3' - $>$ 3 человек
   1. Pstatus - как проживают родители (бинарное значение):  
       - 'T' - living together - живут вместе
       - 'A' - apart - живут раздельно
   1. Medu - оброзование матери (категориальное значение):
       - 0 - нет 
       - 1 - начальное образование (закончила 4 класса)
       - 2 – закончила от 5 до 9 классов 
       - 3 – среднее образование 
       - 4 – высшее образование
   1. Fedu - оброзование отца (категориальное значение): 
       - 0 - нет 
       - 1 - начальное образование (закончил 4 класса)
       - 2 – закончил от 5 до 9 классов 
       - 3 – среднее образование 
       - 4 – высшее образование
   1. Mjob - работа матери (категориальное значение):
       - 'teacher'
       - 'health' - care related
       - 'services' - administrative or police 
       - 'at_home' 
       - 'other'
   1. Fjob - работа отца (категориальное значение):
       - 'teacher'
       - 'health' - care related
       - 'services' - administrative or police 
       - 'at_home' 
       - 'other'
   1. reason - причина выбора *этой* школы (категориальное значение):
       - close to 'home'
       - school 'reputation'
       - 'course' preference 
       - 'other'
   1. guardian - опекун ученика (категориальное значение):
       - 'mother'
       - 'father' 
       - 'other'
   1. traveltime - время пути от школы до дома (категориальное значение):
       - 1 - $<$ 15 минут 
       - 2 - от 15 до 30 минут
       - 3 - от 30 минут до 1 часа
       - 4 -  $>$ 1 часа
   1. studytime - время в неделю, посвяшённое учёбе (категориальное значение):
       - 1 - $<$ 2 часов 
       - 2 - от 2 до 5 часов 
       - 3 - от 5 до 10 часов 
       - 4 - $>$ 10 часов
   1. failures - сколько раз ученик оставался на второй год (категориальное значение):
       - 1 - 1 раз
       - 2 - 2 раза
       - 3 - 3 раза
       - 4 - больше 3 раз 
   1. schoolsup - ученик имеет дополнительную образовательную поддержку (бинарное значение):
       - yes - да
       - no - нет
   1. famsup - ученик получает помощь в образовании со стороны семьи (бинарное значение):
       - yes - да
       - no - нет
   1. paid - ученик посещает дополнительные платные занятия по предмету курса (Math or Portuguese) (бинарное значение):
       - yes - да
       - no - нет
   1. activities - участвует во внекласснойдеятельности (бинарное значение):
       - yes - да
       - no - нет
   1. nursery - посещал детский сад (бинарное значение):
       - yes - да
       - no - нет
   1. higher - хочет получить высшее образование (бинарное значение):
       - yes - да
       - no - нет
   1. internet - имеет доступ к интернету дома (бинарное значение):
       - yes - да
       - no - нет
   1. romantic - находится в романтических отношениях (бинарное значение):
       - yes - да
       - no - нет
   1. famrel - качество отношений с семьёй (категориальное значение) от 1 до 5, где 
       - 1 - very bad 
       - 5 - excellent
   1. freetime - свободное время после школы (категориальное значение) от 1 до 5, где  
       - 1 - very low 
       - 5 - very high
   1. goout - прогулки с друзьями (категориальное значение) от 1 до 5, где 
       - 1 - very low 
       - 5 - very high)
   1. Dalc - потребление алкоголя в будний день (категориальное значение) от 1 до 5, где  
       - 1 - very low 
       - 5 - very high)
   1. Walc - потребление алкоголя в выходные дни (категориальное значение) от 1 до 5, где 
       - 1 - very low
       - 5 - very high)
   1. health - текущее состояние здоровья (категориальное значение) от 1 до 5, где 
        - 1 - very bad
        - 5 - very good)
   1. absences - количество прогулов в школе (числовое значение от 0 до 93)
   1. G1 - оценка за первый период (числовое значение от 0 до 20) 
   1. G2 - оценка за второй период (числовое значение от 0 до 20) 
   1. G3 - итоговая оценка (числовое значение от 0 до 20)

## 2. Сколько в датасете категориальных признаков, какие именно?
---

Из 33 признаково, перечисленных выше, категориальными являются:

1. Medu - оброзование матери
1. Fedu - оброзование отца
1. Mjob - работа матери
1. Fjob - работа отца
1. reason - причина выбора *этой* школы
1. guardian - опекун ученика
1. traveltime - время пути от школы до дома
1. studytime - время в неделю, посвяшённое учёбе
1. failures - сколько раз ученик оставался на второй год
1. famrel - качество отношений с семьёй
1. freetime - свободное время после школы
1. goout - прогулки с друзьями
1. Dalc - потребление алкоголя в будний день
1. Walc - потребление алкоголя в выходные дни
1. health - текущее состояние здоровья

Итого, **15 признаков**.    
К категориальным признакам можно также *условно* отнести **age** и **G1, G2, G3**, поскольку они принимают достаточно ограниченное число значений, 
однако в рамках нашей работы мы будем считать их числовыми значениями.

## 3. Столбец с макимальным количеством уникальных значений категориального признака?
---

In [7]:
categorial = ['Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health']
for column in df.columns:
    if column in categorial:
        #print(df[column].value_counts())
        print(column, df[column].nunique())

Medu 5
Fedu 5
Mjob 5
Fjob 5
reason 4
guardian 3
traveltime 4
studytime 4
failures 4
famrel 5
freetime 5
goout 5
Dalc 5
Walc 5
health 5


Из описания признаков, приведённого в пункте 1, и из результатов работы функции `nunique()` видно, максимальное количество уникальных значений = **5**. Оно достигается у столбцов **Medu, Fedu, Mjob, Fjob, famrel, freetime, goout, Dalc, Walc, health** 

## 4. Есть ли бинарные признаки?
***

In [8]:
i = 1
for column in df.columns:
    if df[column].nunique() == 2:
        print(i, column)
        i += 1

1 school
2 sex
3 address
4 famsize
5 Pstatus
6 schoolsup
7 famsup
8 paid
9 activities
10 nursery
11 higher
12 internet
13 romantic


Из описания признаков, приведённого в пункте 1 и из результатов работы функции, выполненной в блоке выше, видно, что бинарными являются 13 признаков:
**school, sex, address, famsize, Pstatus, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic**

## 5. Какие числовые признаки?
***

In [9]:
i = 1
for column in df.columns:
    if df[column].nunique() > 5:
        print(i, column, df[column].nunique())
        i += 1

1 age 8
2 absences 35
3 G1 18
4 G2 17
5 G3 19


Из описания признаков, приведённого в пункте 1 и из результатов работы функции, выполненной в блоке выше, видно, что *чисто* числовыми являются **age, absences, G1, G2, G3**. Помимо этого, часть категориальных признаков, таких как **famrel, Dalc**, принимающих значения в узком интервале, тоже можно *условно* отнести к числовым, но в рамках нашей работы мы отнесём их к категориальным. 

## 6. Есть ли пропуски?
***

In [10]:
df.isnull().sum(axis = 0)

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

Как видно из результата работы функции `isnull().sum(axis = 0)`, **пропусков в датасете нет**.

## 7. Сколько объектов с пропусками?
***

**0**. *(Смотри пункт 6).*

## 8. Столбец с максимальным количеством пропусков?
***

**Такого столбца нет**. *(Смотри пункт 6).*

## 9. Есть ли на ваш взгляд выбросы, аномальные значения?
***

In [11]:
for column in df.columns:
    print(df[column].value_counts())

GP    772
MS    272
Name: school, dtype: int64
F    591
M    453
Name: sex, dtype: int64
16    281
17    277
18    222
15    194
19     56
20      9
21      3
22      2
Name: age, dtype: int64
U    759
R    285
Name: address, dtype: int64
GT3    738
LE3    306
Name: famsize, dtype: int64
T    923
A    121
Name: Pstatus, dtype: int64
4    306
2    289
3    238
1    202
0      9
Name: Medu, dtype: int64
2    324
1    256
3    231
4    224
0      9
Name: Fedu, dtype: int64
other       399
services    239
at_home     194
teacher     130
health       82
Name: Mjob, dtype: int64
other       584
services    292
teacher      65
at_home      62
health       41
Name: Fjob, dtype: int64
course        430
home          258
reputation    248
other         108
Name: reason, dtype: int64
mother    728
father    243
other      73
Name: guardian, dtype: int64
1    623
2    320
3     77
4     24
Name: traveltime, dtype: int64
2    503
1    317
3    162
4     62
Name: studytime, dtype: int64
0    861
1  

По результам работы функции, выполненной в блоке выше, видим, что все значения в рамках нормы, **выбросов и аномальных значений нет**.

## 10. Столбец с максимальным средним значением после нормировки признаков через стандартное отклонение?
***

Нормируем признак (для числовых признаков) через стандартное отклонение:

In [12]:
import math
numeric = ['age', 'absences', 'G1', 'G2', 'G3']
for column in numeric:
    #print('before:\n', df[column])
    df[column] = (df[column] - df[column].mean())/(math.sqrt(df[column].var()))
    #print('after:\n', df[column])

Также преобразуем текстовые значения в числовые:

In [13]:
texted = ['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob', 'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic']
for column in texted:
    cur_set = set(df[column])
    i = 0
    for item in cur_set:
        df[column] = df[column].replace(item, i)
        i = i + 1

In [18]:
df

school  sex       age  address  famsize  Pstatus  Medu  Fedu  Mjob  \
0          0    0  1.027397        1        0        1     4     4     2   
1          0    0  0.220929        1        0        0     1     1     2   
2          0    0 -1.392007        1        1        0     1     1     2   
3          0    0 -1.392007        1        0        0     4     2     1   
4          0    0 -0.585539        1        0        0     3     3     3   
...      ...  ...       ...      ...      ...      ...   ...   ...   ...   
1039       1    1  2.640333        1        1        1     2     2     4   
1040       1    1  0.220929        1        1        0     3     1     4   
1041       1    1  3.446801        0        0        0     1     1     3   
1042       1    1  1.027397        0        1        0     3     2     4   
1043       1    1  1.833865        1        1        0     1     1     3   

      Fjob  ...  famrel  freetime  goout  Dalc  Walc  health  absences  \
0        0  ...       4         3      4     1     1       3 -0.070027   
1        3  ...       5         3      3     1     1       3 -0.392087   
2        3  ...       4         3      2     2     3       3  0.252034   
3        4  ...       3         2      2     1     1       5 -0.714147   
4        3  ...       4         3      2     1     2       5 -0.714147   
...    ...  ...     ...       ...    ...   ...   ...     ...       ...   
1039     4  ...       5         5      4     4     5       4  1.057185   
1040     4  ...       2         4      5     3     4       2 -0.231057   
1041     3  ...       5         5      3     3     3       3 -0.231057   
1042     3  ...       4         4      1     3     4       5 -0.714147   
1043     2  ...       3         2      3     3     3       5  0.091004   

            G1        G2        G3  
0    -3.758673 -0.074936 -0.088479  
1    -0.741974 -0.074936 -0.088479  
2     0.263592  0.533879  0.170267  
3     0.933969  0.838287  0.687758  
4    -0.071597  0.533879  0.429013  
...        ...       ...       ...  
1039 -0.741974 -0.683750 -0.605971  
1040  0.933969  1.447102  1.205250  
1041 -0.406786 -0.988158 -1.123463  
1042 -0.071597  0.229472 -0.347225  
1043 -1.077163 -0.683750 -0.605971  

[1044 rows x 33 columns]

Найдем максимальное среднее значение (среди столбов с числовым значением):

In [14]:
max_average = -math.inf
max_name = ''
for column in numeric:
    if df[column].mean() > max_average:
        max_name = column
        max_average = df[column].mean()
print(f"Максимальное среднее значение {max_average} имеет столбец {max_name}")

Максимальное среднее значение 1.7695508744983337e-16 имеет столбец age


По результам работы функции, выполненной в блоке выше, видим, что столбец с максимальным средним значением после нормировки признаков - **age**.

## 11. Столбец с целевым признаком?
***

Поскольку мы собераемся делать прогноз итогового балла учащегося, то целевой признак - **G3**.

## 12. Сколько объектов попадает в тренировочную выборку при использовании train_test_split с параметрами test_size = 0.3, random_state = 42?
***

Сначала исключим столбец с целевым признаком из датасета для тренировки:

In [15]:
df_for_training = df.drop('G3', axis=1)
df_for_validation = df['G3'] # df_for_validation = df.loc[:, 'G3']

Разделим данные на тренировочные и проверочные

In [16]:
from sklearn.model_selection import train_test_split
x_train, x_validation, y_train, y_validation = train_test_split(df_for_training, df_for_validation, test_size=.3, random_state=42)

In [17]:
x_train

school  sex       age  address  famsize  Pstatus  Medu  Fedu  Mjob  Fjob  \
447       1    1  0.220929        0        0        0     2     1     3     3   
694       0    0 -1.392007        1        1        1     4     3     3     3   
448       1    0 -0.585539        0        0        0     4     4     0     0   
934       0    1  0.220929        1        0        0     1     1     3     3   
582       1    0  1.833865        1        1        1     1     1     2     3   
..      ...  ...       ...      ...      ...      ...   ...   ...   ...   ...   
87        0    0 -1.392007        1        0        0     4     2     3     3   
330       0    0  1.027397        1        0        0     2     1     3     3   
466       1    1 -1.392007        1        0        0     3     1     3     4   
121       0    1 -1.392007        1        0        0     2     2     4     4   
860       0    1  0.220929        1        1        0     4     4     4     3   

     ...  romantic  famrel  freetime  goout  Dalc  Walc  health  absences  \
447  ...         1       5         5      5     5     5       3  0.574094   
694  ...         1       5         2      2     1     1       5  0.574094   
448  ...         1       4         2      2     1     1       4  0.252034   
934  ...         0       4         3      3     1     2       4 -0.392087   
582  ...         0       1         4      4     1     1       5 -0.714147   
..   ...       ...     ...       ...    ...   ...   ...     ...       ...   
87   ...         0       5         3      3     1     3       1 -0.070027   
330  ...         1       4         3      1     1     1       5  0.896154   
466  ...         0       3         2      3     1     3       4 -0.714147   
121  ...         0       5         5      4     1     2       5  0.252034   
860  ...         1       5         3      5     4     5       3  1.379245   

           G1        G2  
447 -1.077163 -0.379343  
694 -1.077163 -0.988158  
448  1.604347  1.447102  
934  0.263592 -0.379343  
582 -1.747540 -0.988158  
..        ...       ...  
87   1.269158  1.142694  
330  0.263592  0.533879  
466 -0.406786 -0.683750  
121  0.933969  0.533879  
860  0.263592  0.229472  

[730 rows x 32 columns]

Как видно из размеров таблицы, в тренировочную выборку поподёт **730 объектов**.

## 13. Между какими признаками наблюдается линейная зависимость (корреляция)?
***

Для выявления линейной зависимости признаков (мультиколлениарности), будем использовать VIF (Variable Inflation Factors). 
Теоритические данные взяты [отсюда](https://www.analyticsvidhya.com/blog/2020/03/what-is-multicollinearity/).

In [28]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif_data = pd.DataFrame()
vif_data["column_name"] = df_for_training.columns
vif_data["VIF"] = [variance_inflation_factor(df_for_training.values, i) for i in range(len(df_for_training.columns))]
#vif_data = vif_data.transpose()

Выведем признаки со значением VIF $\geq 10$:

In [67]:
vif_data.loc[vif_data['VIF'] >= 10.0]

column_name        VIF
6         Medu  12.938283
7         Fedu  10.457012
9         Fjob  10.354606
23      famrel  18.215319
24    freetime  12.421484
25       goout  12.022801

Как мы видим, наибольший ($>10$ ) показатель VIF у **Medu, Fedu, Fjob, famrel, freetime, goout**.

## 14. Сколько признаков достаточно для объяснения 90% дисперсии после применения метода PCA?
***

### Тут надо бы вспомнить метод РСА

## 15. Какой признак вносит наибольший вклад в первую компоненту?
***

### ??? Что такое первая компонента ???

## Доп. задание: построить двухмерное представление данных с помощью алгоритма tSNE. На сколько кластеров визуально на ваш взгляд разделяется выборка?
***